File: array/SlidingWindow/fixed_size/cpp/SlidingWindow.cpp
#include "SlidingWindow.hpp"

using namespace std;

namespace {
bool CheckedAdd(long long a, long long b, long long& result) {
    if ((b > 0 && a > 9223372036854775807LL - b) ||
        (b < 0 && a < (-9223372036854775807LL - 1LL) - b)) {
        return false;
    }
    result = a + b;
    return true;
}

bool CheckedSub(long long a, long long b, long long& result) {
    if ((b < 0 && a > 9223372036854775807LL + b) ||
        (b > 0 && a < (-9223372036854775807LL - 1LL) + b)) {
        return false;
    }
    result = a - b;
    return true;
}
}  // namespace

bool BestFixedWindow(const vector<long long>& values, int k, long long& bestSum, int& bestLeft, int& bestRight) {
    long long windowSum = 0;
    for (int i = 0; i < k; ++i) {
        if (!CheckedAdd(windowSum, values[i], windowSum)) {
            return false;
        }
    }

    bestSum = windowSum;
    bestLeft = 0;
    bestRight = k - 1;

    for (int right = k; right < static_cast<int>(values.size()); ++right) {
        if (!CheckedSub(windowSum, values[right - k], windowSum)) {
            return false;
        }
        if (!CheckedAdd(windowSum, values[right], windowSum)) {
            return false;
        }
        const int left = right - k + 1;
        if (windowSum > bestSum) {
            bestSum = windowSum;
            bestLeft = left;
            bestRight = right;
        }
    }

    return true;
}

File: array/SlidingWindow/fixed_size/java/SlidingWindow.java
public final class SlidingWindow {
    private SlidingWindow() {}

    private static boolean addOverflow(long a, long b) {
        return (b > 0 && a > Long.MAX_VALUE - b) || (b < 0 && a < Long.MIN_VALUE - b);
    }

    private static boolean subOverflow(long a, long b) {
        return (b < 0 && a > Long.MAX_VALUE + b) || (b > 0 && a < Long.MIN_VALUE + b);
    }

    public static Result bestFixedWindow(long[] values, int k) {
        long windowSum = 0;
        for (int index = 0; index < k; index++) {
            if (addOverflow(windowSum, values[index])) {
                return null;
            }
            windowSum += values[index];
        }

        long bestSum = windowSum;
        int bestLeft = 0;
        int bestRight = k - 1;

        for (int right = k; right < values.length; right++) {
            if (subOverflow(windowSum, values[right - k])) {
                return null;
            }
            windowSum -= values[right - k];
            if (addOverflow(windowSum, values[right])) {
                return null;
            }
            windowSum += values[right];

            int left = right - k + 1;
            if (windowSum > bestSum) {
                bestSum = windowSum;
                bestLeft = left;
                bestRight = right;
            }
        }

        return new Result(bestSum, bestLeft, bestRight);
    }

    public record Result(long bestSum, int bestLeft, int bestRight) {}
}
